### PR TITLE
Charlesmchen/blocking7

### PR DESCRIPTION
**Pull request by charlesmchen**
* Block actions in message view for blocked contact conversations.
* Block actions in message view for blocked group conversations.
* Add block status indicator view.

PTAL @michaelkirk 

screen shot 2017-04-04 at 3 36 48 pm
screen shot 2017-04-04 at 3 36 34 pm

### CODE REVIEW

**Comment by charlesmchen** [comment]
Discussion: 

* It wasn't 100% clear what to do when user taps the "group has blocked members indicator".  I chose to present the "block list" view. It won't indicate which users in the block list are in the group, but I don't think it's worth it to build a view that would do so and most people's block lists should be quite short.
* The "1:1" and "group" block indicators "float" about the conversation view and obscure some of its contents.  Most critically, it obscures the "load more/older messages" interaction.  I think we should discuss how to present this.

**Comment by charlesmchen** [comment]
PTAL @michaelkirk

File: Signal/src/ViewControllers/BlockListUIUtils.m [hunk 1]
```diff
@@ -27,10 +29,12 @@ + (void)showBlockContactActionSheet:(Contact *)contact
     if (phoneNumbers.count < 1) {
         DDLogError(@"%@ Contact has no phone numbers", self.tag);
         OWSAssert(0);
-        [self showBlockFailedAlert:fromViewController];
-        if (completionBlock) {
-            completionBlock(NO);
-        }
+        [self showBlockFailedAlert:fromViewController
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/BlockListUIUtils.m]: Not calling the completion block until after the confirmation/failure alerts have been dismissed, allows us to present alerts/vcs in the completion handler.

File: Signal/src/ViewControllers/MessagesViewController.m [hunk 2]
```diff
@@ -500,6 +524,133 @@ - (void)viewWillAppear:(BOOL)animated
...
+    }
+
+    NSString *blockStateMessage = nil;
+    if ([self isBlockedContactConversation]) {
+        blockStateMessage = NSLocalizedString(@"MESSAGES_VIEW_CONTACT_BLOCKED",
+                                              @"Indicates that this 1:1 conversation has been blocked.");
+    } else {
+        int blockedGroupMemberCount = [self blockedGroupMemberCount];
+        if (blockedGroupMemberCount == 1) {
+            blockStateMessage = [NSString stringWithFormat:NSLocalizedString(@"MESSAGES_VIEW_GROUP_1_MEMBER_BLOCKED_FORMAT",
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: This doesn't seem like a "format" string, why are we passing in count?

File: Signal/src/ViewControllers/MessagesViewController.m [hunk 3]
```diff
@@ -500,6 +524,133 @@ - (void)viewWillAppear:(BOOL)animated
...
+                                                                             @"Indicates that a single member of this group has been blocked."),
+                                 blockedGroupMemberCount];
+        } else if (blockedGroupMemberCount > 1) {
+            blockStateMessage = [NSString stringWithFormat:NSLocalizedString(@"MESSAGES_VIEW_GROUP_N_MEMBERS_BLOCKED_FORMAT",
+                                                                             @"Indicates that some members of this group has been blocked. Embeds "
+                                                                             @"{{the number of blocked user in this group}}."),
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: typo: number of blocked user**s** in this group

File: Signal/src/ViewControllers/MessagesViewController.m [hunk 4]
```diff
@@ -500,6 +524,133 @@ - (void)viewWillAppear:(BOOL)animated
...
+        return NO;
+    }
+    NSString *contactIdentifier = ((TSContactThread *)self.thread).contactIdentifier;
+    return [[_blockingManager blockedPhoneNumbers] containsObject:contactIdentifier];
+}
+
+- (int)blockedGroupMemberCount
+{
+    if (![self.thread isKindOfClass:[TSGroupThread class]]) {
+        return 0;
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: nit:I think `OWSAssert(NO); `

File: Signal/src/ViewControllers/MessagesViewController.m [hunk 5]
```diff
@@ -500,6 +524,133 @@ - (void)viewWillAppear:(BOOL)animated
...
+    }
+    
+    if (blockStateMessage) {
+        UILabel *label = [UILabel new];
+        label.font = [UIFont ows_mediumFontWithSize:14.f];
+        label.text = blockStateMessage;
+        label.textColor = [UIColor whiteColor];
+        
+        UIView * blockStateIndicator = [UIView new];
+        blockStateIndicator.backgroundColor = [UIColor ows_signalBrandBlueColor];
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: I think we want material blue. signalBrandBlueColor is a bit darker and works well for text. e.g. compare to the navbar background.

File: Signal/translations/en.lproj/Localizable.strings [hunk 6]
```diff
@@ -529,6 +529,15 @@
 /* message footer while attachment is uploading */
 "MESSAGE_STATUS_UPLOADING" = "Uploading...";
 
+/* Indicates that this 1:1 conversation has been blocked. */
+"MESSAGES_VIEW_CONTACT_BLOCKED" = "This User is Blocked";
```
Comment (michaelkirk) [inline Signal/translations/en.lproj/Localizable.strings]: Maybe "You Blocked this User" to be clear that the user isn't e.g. banned from the signal service.

File: Signal/translations/en.lproj/Localizable.strings [hunk 7]
```diff
@@ -529,6 +529,15 @@
...
+
+/* Indicates that a single member of this group has been blocked. */
+"MESSAGES_VIEW_GROUP_1_MEMBER_BLOCKED_FORMAT" = "1 Blocked Member of this Group";
```
Comment (michaelkirk) [inline Signal/translations/en.lproj/Localizable.strings]: Maybe: Similarly, maybe "Blocked 1 Member of this Group" and "Blocked %d Members of this Group".

take it or leave it.

File: Signal/src/ViewControllers/MessagesViewController.m (hunk 2 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessagesViewController.m]: Done.

File: Signal/src/ViewControllers/MessagesViewController.m (hunk 3 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessagesViewController.m]: Done.

File: Signal/src/ViewControllers/MessagesViewController.m (hunk 5 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessagesViewController.m]: I'm thinking we should use a color that is more "alert-y", somewhere on the red, orange, yellow spectrum - to draw the eye and to emphasize that some functionality is disabled.

Here's `[UIColor ows_redColor]`:

screen shot 2017-04-05 at 11 29 12 am


WDYT?

File: Signal/src/ViewControllers/MessagesViewController.m (hunk 4 above)
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessagesViewController.m]: Made a few changes here including adding asserts.

File: Signal/translations/en.lproj/Localizable.strings (hunk 6 above)
Comment (charlesmchen) [inline Signal/translations/en.lproj/Localizable.strings]: That's a really good idea. Done.

File: Signal/translations/en.lproj/Localizable.strings (hunk 7 above)
Comment (charlesmchen) [inline Signal/translations/en.lproj/Localizable.strings]: Yup.

screen shot 2017-04-05 at 11 34 27 am

File: Signal/src/ViewControllers/MessagesViewController.m (hunk 5 above)
Comment (michaelkirk) [inline Signal/src/ViewControllers/MessagesViewController.m]: :+1:

**Comment by charlesmchen** [comment]
> Is it going to be clear enough to the user why we're popping the action sheet? E.g. we could be really explicit.

I had a different idea, since I tend to think long alert titles look cramped.  When we show these "you need to unblock to do that" alert, I now re-show the "block state indicator".

screen shot 2017-04-05 at 11 29 07 am

It clearly indicates the state (in a consistent way), it's presentation draws the eye to remind the user, and it offers the user a separate means to resolve the issue if the cancel.

WDYT?

**Comment by charlesmchen** [comment]
> One issue I noticed is that if I block Alice, then message Alice, I see the "Unblock Alice?" dialog, as expected. When I accept, the message sends as expected, but when the action sheet dismisses, they text-input is not visible. (I think it's behind the keyboard).

Fixed.  I also avoided some keyboard animations around these "unblock" action sheets.

**Pull request merged**